src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -72,7 +72,13 @@ internal class DefaultHealthCheckService : HealthCheckService
                         HealthReportEntry entry;
                         try
                         {
-                            var result = await healthCheck.CheckHealthAsync(context, cancellationToken);
+                            HealthCheckResult result;
+                            using (var timeoutCancellationTokenSource = new CancellationTokenSource(registration.Timeout))
+                            using (cancellationToken.Register(() => timeoutCancellationTokenSource.Cancel()))
```

Comment:
You can use https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.createlinkedtokensource?view=netframework-4.7.2 to combine two tokens.

Reply:
Thanks for the tip! Didn't know this method, very useful !